cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,424 @@
+      // "identity operator" is used
+      return this->IDENTITY;
+    } else {
+      return {beta, 1.0};
+    }
+  }
+};
+
+template <typename T>
+class ewma_adjust_denominator_nonull_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_adjust_denominator_nonull_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(pair_type<T> pairs) { return {beta, 1.0}; }
+};
+
+template <typename T>
+class ewma_noadjust_nonull_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_noadjust_nonull_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(thrust::tuple<T, size_type> data)
+  {
+    T const beta          = this->beta;
+    T const input         = thrust::get<0>(data);
+    size_type const index = thrust::get<1>(data);
+    if (index == 0) {
+      return {beta, input};
+    } else {
+      return {beta, (1.0 - beta) * input};
+    }
+  }
+};
+
+template <typename T>
+class ewma_noadjust_null_functor : public ewma_functor_base<T> {
+ private:
+  T beta;
+
+ public:
+  ewma_noadjust_null_functor(T beta) : beta{beta} {}
+
+  __device__ pair_type<T> operator()(thrust::tuple<T, size_type, bool, size_type> data)
+  {
+    T input           = thrust::get<0>(data);
+    size_type index   = thrust::get<1>(data);
+    bool is_valid     = thrust::get<2>(data);
+    size_type nullcnt = thrust::get<3>(data);
+
+    T beta = this->beta;
+
+    if (index == 0) {
+      return {beta, input};
+    } else {
+      if (is_valid and nullcnt == 0) {
+        // preceeding value is valid, return normal pair
+        return {beta, (1.0 - beta) * input};
+      } else if (is_valid and nullcnt != 0) {
+        // one or more preceeding values is null, adjust by how many
+        T factor = (1.0 - beta) + pow(beta, nullcnt + 1);
+        return {(beta * (pow(beta, nullcnt)) / factor), ((1.0 - beta) * input) / factor};
+      } else {
+        // value is not valid
+        return this->IDENTITY;
+      }
+    }
+  }
+};
+
+/**
+* @brief Return an array whose values y_i are the number of null entries
+* in between the last valid entry of the input and the current index.
+* Example: {1, NULL, 3, 4, NULL, NULL, 7}
+        -> {0, 0     1, 0, 0,    1,    2}
+*/
+rmm::device_uvector<cudf::size_type> null_roll_up(column_view const& input,
+                                                  rmm::cuda_stream_view stream)
+{
+  rmm::device_uvector<cudf::size_type> output(input.size(), stream);
+
+  auto device_view = column_device_view::create(input);
+  auto valid_it    = cudf::detail::make_validity_iterator(*device_view);
+
+  // Invert the null iterator
+  thrust::transform(rmm::exec_policy(stream),
+                    valid_it,
+                    valid_it + input.size(),
+                    output.begin(),
+                    [] __device__(bool valid) -> bool { return 1 - valid; });
+
+  // 0, 1, 0, 1, 1, 0 -> 0, 0, 1, 0, 0, 2
```

Comment:
I don't think inclusive_scan_by_key is doing what this comment says since you are using the nulls as the keys and the values:
// keys: 0, 1, 0, 1, 1, 0
// values: 0, 1, 0, 1, 1, 0
// result: 0, 1, 0, 1, 2, 0